### PR TITLE
Comment analysis by keyword search

### PR DESCRIPTION
**Pull request by olu144**
keywordsearch - Edited

Users can now input a keyword to have it's top comments analyzed as opposed to inputting a username or channel id. In the example above "Java Basics" was the keyword and the top comments from the search were analyzed.
There has also been changes in variable names to improve readability

### CODE REVIEW

File: capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import java.io.BufferedReader;
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import org.json.simple.JSONObject;    
+import java.util.ArrayList;
+import java.util.Arrays;
+
+/** Servlet that returns youtube api data. */
```
Comment (chrismohrii) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java]: nit: Let's update this comment & on other servlets if necessary

File: capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+@WebServlet("/keyword_search_servlet")
+public class YoutubeKeywordServlet extends HttpServlet {
+  private static final String BASE_URL = "https://www.googleapis.com/youtube/v3/search?part=snippet";
+  private static final String KEY = "API_KEY";
+  private static final String NUM_RESULTS = "5";
+  private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
+  OkHttpClient client = new OkHttpClient();
+  @Override
```
Comment (chrismohrii) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java:44]: nit: Let's put a separating new line for line 42

**Review by chrismohrii — COMMENTED** [review]
Looks good. 

The keyword and trending servlet are almost identical - if possible, it might be a good idea to combine them or find another way to avoid code repetition

File: capstone/src/main/webapp/youtubeScript.js [hunk 3]
```diff
@@ -86,28 +87,44 @@ async function inputCommentsToPerspective(commentsList) {
...
+/** Returns a map of attribute score sums from an array of JSON's */
+function getAttributeTotals(attributeScores) {
+  const requestedAttributes = getRequestedAttributes();
+  const attributeTotals = new Map();    
+  for (let i = 0; i < requestedAttributes.length; i++) {
+    for (let j = 0; j < attributeScores.length; j++) {
+      if (attributeTotals.has(requestedAttributes[i])) {
+        attributeTotals.set(requestedAttributes[i], attributeTotals.get(requestedAttributes[i]) + attributeScores[j].attributeScores[requestedAttributes[i]].summaryScore.value);
+      } else {
+        attributeTotals.set(requestedAttributes[i], attributeScores[j].attributeScores[requestedAttributes[i]].summaryScore.value);
```
Comment (danieldortega77) [inline capstone/src/main/webapp/youtubeScript.js]: Same note as [this comment](https://github.com/googleinterns/step40-2020/pull/29#discussion_r459720932), we might want to create local variables to increase readability here

**Review by danieldortega77 — COMMENTED** [review]
Same general comment as Chris  - maybe we could specify what we want to search with parameters to the Get request to combine the two servlets.

File: capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java (hunk 1 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java]: addressed in "address Daniel's and Chris's nits"

File: capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java (hunk 2 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java:44]: fix override spacing

File: capstone/src/main/webapp/youtubeScript.js (hunk 3 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "formatting improvements"

**Review by danieldortega77 — APPROVED** [review]
LGTM

**Review by chrismohrii — APPROVED** [review]
LGTM

File: capstone/src/main/webapp/youtubeScript.js [hunk 4]
```diff
@@ -86,28 +85,47 @@ async function inputCommentsToPerspective(commentsList) {
...
     }
-    for (var i = 0; i < requestedAttributes.length; i++) {
-      for (var j = 0; j < attributeScores.length; j++) {
-        if (attributeTotals.has(requestedAttributes[i])) {
-          attributeTotals.set(requestedAttributes[i], attributeTotals.get(requestedAttributes[i]) + attributeScores[j].attributeScores[requestedAttributes[i]].summaryScore.value);
-        } else {
-          attributeTotals.set(requestedAttributes[i], attributeScores[j].attributeScores[requestedAttributes[i]].summaryScore.value);
-        }
+  }
+  let totalNumberOfComments = commentsList[0].items.length * commentsList.length;
```
Comment (dslucas) [inline capstone/src/main/webapp/youtubeScript.js]: Just curious, are all the `items` in each entry of `commentsList` expected to be the same length?

File: capstone/src/main/webapp/youtubeScript.js [hunk 5]
```diff
@@ -264,9 +286,66 @@ function showCategories() {
...
-    container.appendChild(document.createTextNode (" "));
+    const categoryContainer = document.getElementById('category-container');
+    categoryContainer.appendChild(radiobox);
+    categoryContainer.appendChild(label);
+    categoryContainer.appendChild(document.createTextNode(" "));
   }
 }
+
+/** Converts perspective results to knoop scale then to mohs*/
+function perspectiveToxicityScale(attributeAverages) {
```
Comment (dslucas) [inline capstone/src/main/webapp/youtubeScript.js]: nit: Let's use a more verb-y name for this (e.g. `scaleAverages()` or `convertToMohsScale()` or something like that).

**Review by dslucas — APPROVED** [review]

File: capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String searchTerm = request.getParameter("searchTerm");
+    String completeUrl = BASE_URL + "&maxResults=" + NUM_RESULTS + "&q=" + searchTerm + "&key=" + KEY;
+    String output = get(completeUrl);
+    response.setContentType("application/json");
+    response.getWriter().println(output);  
+  }
+  
+  /** Makes a GET request. */
+  private String get(String url) throws IOException {
```
Comment (dborkan) [inline capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java]: I think this function exists now in a few different places - is it possible to refactor this code so it's only defined in 1 .java file?

File: capstone/src/main/webapp/youtubeScript.js [hunk 7]
```diff
@@ -21,13 +21,13 @@ const ATTRIBUTES_BY_LANGUAGE = {
...
 };
 
+/** Category names are mapped to youtube category numbers */
 const YOUTUBE_CATEGORIES = {
   'Autos&Vehicles': 2,
   'Comedy': 23,
   'Entertainment': 24,
   'Film&Animation': 1,
   'Gaming': 20,
-  'How-to&Style' : 26,
```
Comment (dborkan) [inline capstone/src/main/webapp/youtubeScript.js:31]: Just curious, any reason why this was removed?

**Review by dborkan — COMMENTED** [review]
New code overall looks good, just 1 comment about possible refactoring.  You might want to include in the description that this includes a lot of variable renaming and changes for `let` and `const`.  Also once #21 is merged you may want to sync or rebase this, just so the diffs only show the new code.

File: capstone/src/main/webapp/youtubeScript.js (hunk 5 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: addressed in "optimize and clarify PerspectiveToxicityScale function"

File: capstone/src/main/webapp/youtubeScript.js (hunk 4 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js]: Yes, the length of the items in each entry of commentsList is dependent on the servlet response. The lengths are expected to be the same unless there is an API error which would have been caught before this point

File: capstone/src/main/webapp/youtubeScript.js (hunk 7 above)
Comment (olu144) [reply capstone/src/main/webapp/youtubeScript.js:31]: I was experiencing issues where results were not consistently returned. I consulted online forums and found others were having similar issues. I plan to continue experimenting to see if issue is resolved.

File: capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java (hunk 6 above)
Comment (olu144) [reply capstone/src/main/java/com/google/sps/servlets/YoutubeKeywordServlet.java]: addressed in "reduce code repetition for get function".

**Review by dborkan — APPROVED** [review]
👍

**Pull request merged**